### PR TITLE
Update Ledger to allow canisters to transfer ICP

### PR DESCRIPTION
**Pull request by roman-kashitsyn**

### CODE REVIEW

File: proposals/network_canister_management/ledger_20211117T2319Z.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
+## Summary
+
+Updating Ledger to Wasm module with hash `61a312e4bf4f86899edabf6b3cedc504b2b0825d9b7956abea65582095692a9a` built from git commit `88d24e7def3b6ef41508d753f4037dd76d940a6d`.
+
+## Rationale
+
+This update contains changes required to allow all canisters to transfer ICP by sending messages to the Ledger canister:
+
+  * The ledger doesn't apply the whitelist when it decides whether to allow the transfer to the caller.
```
Comment (JensGroth) [inline proposals/network_canister_management/ledger_20211117T2319Z.md:9]:
```suggestion
  * The ledger doesn't apply the whitelist when it decides whether to allow the caller to make a transfer.
```

File: proposals/network_canister_management/ledger_20211117T2319Z.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    All principals except for the anonymous are allowed to transfer ICP now.
```
Comment (JensGroth) [inline proposals/network_canister_management/ledger_20211117T2319Z.md:10]:
```suggestion
    All principals except for the anonymous principal are allowed to transfer ICP now.
```

File: proposals/network_canister_management/ledger_20211117T2319Z.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+The newly exposed candid interface:
+
+```candid
+// Amount of ICP tokens, measured in 10^-8 of a token.
+type ICP = record {
+     e8s : nat64;
+};
+
+// Number of nanoseconds from the UNIX epoch in UTC timezone.
```
Comment (JensGroth) [inline proposals/network_canister_management/ledger_20211117T2319Z.md:23]:
```suggestion
// Number of nanoseconds from the UNIX epoch (00:00:00 UTC, Jan 1, 1970).
```

File: proposals/network_canister_management/ledger_20211117T2319Z.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    amount: ICP;
+    // The amount that the caller pays for the transaction.
+    // Must be 10000 e8s.
+    fee: ICP;
+    // The subaccount from which the caller wants to transfer funds.
+    // If null, the ledger uses the default (all zeros) subaccount to compute the source address.
+    // See comments for the `SubAccount` type.
+    from_subaccount: opt SubAccount;
+    // The destination account.
+    // If the transfer is successful, the balance of this address increases by `amount`.
```
Comment (JensGroth) [inline proposals/network_canister_management/ledger_20211117T2319Z.md:59]: The two lines above are a bit confusing. Generally: are we calling it account, account identifier or address?

File: proposals/network_canister_management/ledger_20211117T2319Z.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    to: AccountIdentifier;
+    // The point in time when the caller created this request.
+    // If null, the ledger uses current IC time as the timestamp.
+    created_at_time: opt Timestamp;
+};
+
+type TransferError = variant {
+    // The fee that the caller specified in the transfer request was not the one that ledger expects.
```
Comment (JensGroth) [inline proposals/network_canister_management/ledger_20211117T2319Z.md:67]:
```suggestion
    // The fee that the caller specified in the transfer request was not the one that the ledger expects.
```

File: proposals/network_canister_management/ledger_20211117T2319Z.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    // The caller can change the transfer fee to the `expected_fee` and retry the request.
+    BadFee : record { expected_fee : ICP; };
+    // The account specified by the caller doesn't have enough funds.
+    InsufficientFunds : record { balance: ICP; };
+    // The request is too old.
+    // The ledger only accepts requests created within 24 hours window.
```
Comment (JensGroth) [inline proposals/network_canister_management/ledger_20211117T2319Z.md:73]:
```suggestion
    // The ledger only accepts requests created within a 24 hours window.
```

File: proposals/network_canister_management/ledger_20211117T2319Z.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    // This is a non-recoverable error.
+    TxTooOld : record { allowed_window_nanos: nat64 };
+    // The caller specified `created_at_time` that is too far in future.
+    // The caller can retry the request later.
+    TxCreatedInFuture : null;
+    // The ledger has already executed the request.
+    // `duplicate_of` field is equal to the index of the block containing the original transaction.
+    TxDuplicate : record { duplicate_of: BlockIndex; }
```
Comment (JensGroth) [inline proposals/network_canister_management/ledger_20211117T2319Z.md:81]:
```suggestion
    TxDuplicate : record { duplicate_of: BlockIndex; };
```

File: proposals/network_canister_management/ledger_20211117T2319Z.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+service : {
+  // Transfers tokens from a subaccount of the caller to the destination address.
+  // The source address is computed from the principal of the caller and the specified subaccount.
+  // When successful, returns the index of the block containing the transaction.
+  transfer : (TransferArgs) -> (TransferResult);
+
+  // Returns the amount of ICP on the specified account.
+  account_balance : (AccountBalanceArgs) -> (ICP) query;
+}
```
Comment (JensGroth) [inline proposals/network_canister_management/ledger_20211117T2319Z.md:102]:
```suggestion
}
```
I'm not familiar with the semicolon policy of candid, do you need one to terminate here?

File: proposals/network_canister_management/ledger_20211117T2319Z.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
+## Summary
+
+Updating Ledger to Wasm module with hash `61a312e4bf4f86899edabf6b3cedc504b2b0825d9b7956abea65582095692a9a` built from git commit `88d24e7def3b6ef41508d753f4037dd76d940a6d`.
+
```
Comment (JensGroth) [inline proposals/network_canister_management/ledger_20211117T2319Z.md:4]:
```suggestion
## Relates to

The proposal implements community motion 20588.
```

File: proposals/network_canister_management/ledger_20211117T2319Z.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+## Rationale
+
+This update contains changes required to allow all canisters to transfer ICP by sending messages to the Ledger canister:
```
Comment (JensGroth) [inline proposals/network_canister_management/ledger_20211117T2319Z.md:7]:
```suggestion
Community motion 20588 requested that all principals including canisters should be able to transfer ICP. This update contains changes required to allow all canisters to transfer ICP by sending messages to the Ledger canister:
```

File: proposals/network_canister_management/ledger_20211117T2319Z.md (hunk 8 above)
Comment (roman-kashitsyn) [reply proposals/network_canister_management/ledger_20211117T2319Z.md:102]: In Candid, semicolon is not a terminator, but a separator. Though it's allowed to end a sequence with a semicolon.

File: proposals/network_canister_management/ledger_20211117T2319Z.md (hunk 4 above)
Comment (roman-kashitsyn) [reply proposals/network_canister_management/ledger_20211117T2319Z.md:59]:
```suggestion
    // If the transfer is successful, the balance of this account increases by `amount`.
```

File: proposals/network_canister_management/ledger_20211117T2319Z.md (hunk 3 above)
Comment (roman-kashitsyn) [reply proposals/network_canister_management/ledger_20211117T2319Z.md:23]: > nanosecs is not real time but IC time

Not all timestamps here are IC time: the transaction `created_at_time` timestamp is chosen by the client, it's not necessary IC time. Furthermore, there is no globally synchronized "IC time", different subnets use different clocks. So it's quite possible that transaction.create_at_time will have a timestamp that is larger than the timestamp of the block it's included into.

Comment (JensGroth) [reply proposals/network_canister_management/ledger_20211117T2319Z.md:23]: Ok, updated suggestion to just specify the epoch (the nitpick being that UNIX epoch is the same in all timezones)

**Review by JensGroth — APPROVED** [review]

**Pull request merged**